File: BaekJoon/15663_N과M(9)_BackTracking.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

// N과M(9) 15663 BackTracking
// 중복 방지가 핵심인 문제
// O( M * N^M )


int main(void)
{
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    cout.tie(nullptr);
    
    int n, m;
    cin >> n >> m;

    vector<int> v(n);
    for (int& a : v)
    {
        cin >> a;
    }

    // 사전순 처리를 위한 오름차순 정렬
    sort(v.begin(), v.end());

    vector<bool> isUsed(n, false);  // 중복 방지
    vector<int> seq;
    seq.reserve(m);

    auto DFS = [&](auto&& self, int depth) -> void {
        // m 개의 수를 모두 선택했다면 출력
        if (depth == m)
        {
            for (const auto& a : seq)
            {
                cout << a << ' ';
            }
            cout << '\n';
            return;
        }

        // 같은 깊이에서 동일 수 선택 방지
        int prev{ -1 };
        for (int i = 0; i < n; ++i)
        {
            // 이미 수열에 있는 수이거나, 이전에 선택한 수면 건너뜀
            if (isUsed[i] || v[i] == prev)
            {
                continue;
            }

            isUsed[i] = true;
            seq.push_back(v[i]);
            self(self, depth + 1);
            seq.pop_back();
            isUsed[i] = false;
            prev = v[i];
        }
    };

    DFS(DFS, 0);


    return 0;
}

File: BaekJoon/Java/15663_N과M(9)_BackTracking.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

// N과M(9) 15663 BackTracking
// 중복 방지가 핵심인 문제
// O( M * N^M )


public class Main {
    static int n, m;
    static int[] nums, seq;
    static boolean[] used;
    static StringBuilder sb = new StringBuilder();

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        n = Integer.parseInt(st.nextToken());
        m = Integer.parseInt(st.nextToken());

        nums = new int[n];
        st = new StringTokenizer(br.readLine());
        for (int i = 0; i < n; ++i) {
            nums[i] = Integer.parseInt(st.nextToken());
        }

        Arrays.sort(nums);
        used = new boolean[n];
        seq = new int[m];

        DFS(0);

        System.out.println(sb);
    }

    private static void DFS(int depth) {
        if (depth == m) {
            for (int i = 0; i < m; ++i) {
                sb.append(seq[i]).append(' ');
            }
            sb.append('\n');
            return;
        }

        int prev = -1;
        for (int i = 0; i < n; ++i) {
            if (used[i] || nums[i] == prev)
                continue;

            used[i] = true;
            seq[depth] = nums[i];
            DFS(depth + 1);
            used[i] = false;
            prev = nums[i];
        }
    }
}
